### ISSUE TITLE
新的paddle 2.3rc与 2.2或2.1版梯度更新不同，导致模型loss NAN

### ISSUE THREAD

**Issue by studyplacefor** [PaddlePaddle/Paddle#42348]
### bug描述 Describe the Bug

  这次是  继 issue #42152 之后的另一坑吧！
 
同一代码，同一输入变量， 只是卷积的初始参数不同。 （我没有看过2.3rc 与 2.2/2.1 版的卷积参数初始化参数是不是真的有什么不同，按理说应该没有的）

----------------
1000 iter 内  loss 递减幅度 (打印/100step )：
2.1版
![image](https://user-images.githubusercontent.com/50259160/165680941-f335f74e-a5e3-4384-8795-f8f83ec9bb61.png)

2.3rc版
![image](https://user-images.githubusercontent.com/50259160/165681143-3a40a22e-4c2c-4436-8fcf-1877ba6d500c.png)

这种行为表现，使得相同的项目代码
        在paddle 2.1 上是正常跑的（遗憾issue #42152）
        在           2.3rc  是loss 超大，好像是数值溢出  且很快loss  <50epoch 左右>  NAN 
 
以下是实际项目的截图：
![image](https://user-images.githubusercontent.com/50259160/165681890-7c16231d-00cf-4082-8750-b12058dc7905.png)
![image](https://user-images.githubusercontent.com/50259160/165682269-cc382bea-6247-4e8b-b979-d03b84a9d02b.png)

### 其他补充信息 Additional Supplementary Information

复现代码：
[xtestpd.py.txt](https://github.com/PaddlePaddle/Paddle/files/8579722/xtestpd.py.txt)

用到的  tensor  .   numpy  
[npy.zip](https://github.com/PaddlePaddle/Paddle/files/8579729/npy.zip)

**Comment by chenwhql** [comment]
您说的是2.3rc那张图中，loss急剧下降的现象吧

我在本地用您提供的脚本和数据，尝试复现了一下，发现2.2和2.1好像也是同样现象，是还需要什么配置才能按预期执行吗？

2.2.2
![image](https://user-images.githubusercontent.com/22561442/165711742-aa469137-1d06-496f-96f4-21677c8685ba.png)


2.1.3
![image](https://user-images.githubusercontent.com/22561442/165711478-a7efe876-f075-4ae9-9159-07866fa4739c.png)

**Comment by studyplacefor** [comment]
@[chenwhql](https://github.com/chenwhql)，多谢你的回复！
我的本地上即安有torch,双安有paddle的！ 
   GTX1070ti  :
![image](https://user-images.githubusercontent.com/50259160/165743915-c7e6a0b2-b352-463a-935c-0dd69dba0b5d.png)

aistudio r2.3.0rc     V100 ：
![image](https://user-images.githubusercontent.com/50259160/165750854-6bcc53d6-3222-4687-8391-892f0ef5c054.png)

测试的代码就是我上传的xtestpd.py   ,  python  xtestpd.py  
**模型的输入就是用  我上传的*.npy 文件**。

aistudio 上用的是这环境：
![image](https://user-images.githubusercontent.com/50259160/165744570-3382e9e8-531c-4439-9a52-b1b308005fc7.png)

我又重新测了一次，还是像我上面描述那样的现象： 2.3.0rc 上求loss，梯度降得极快，以至于在实际项目时，几轮就NAN,而2.1.2上则正常，然而却有 后向求导的耗时问题（和torch比，差得太远）

**Comment by qtz980805** [comment]
确实，使用2.3.0rc我也出现几次迭代就NAN的情况了，之前在2.2.0上Loss是正常的。

**Comment by chenwhql** [comment]
我这边本地之前没有测试2.1.2，所以我又测试了一下，但还是没有得到预期效果，测试如下：
![image](https://user-images.githubusercontent.com/22561442/166934985-459f6568-1a13-4f9f-8297-3b7a76d7a2be.png)

和您本地环境的差别的话，一个是Python版本不同，一个是Driver API Version不同？

您在aistudio上测试2.2或者2.1的是正常的吗？

**Comment by chenwhql** [comment]
测试代码应该没有改动，您check下？

```python
import paddle
import paddle.nn as nn
import paddle.nn.functional as F
import paddle.optimizer as optim

import parser
import argparse

class CausalConv1d(nn.Conv1D):
    def __init__(self, in_channels, out_channels, kernel_size=2, dilation=1, **kwargs):
        super(CausalConv1d, self).__init__(
            in_channels,
            out_channels,
            kernel_size,
            padding=dilation*(kernel_size-1),
            dilation=dilation,
            **kwargs)
        
    def forward(self, input):
        out = super(CausalConv1d, self).forward(input)
        return out[:, :, :-self._padding]

class WavenetLayer(nn.Layer):
    def __init__(self, residual_channels, skip_channels, cond_channels, kernel_size=2, dilation=1):
        super(WavenetLayer, self).__init__()

        self.causal = CausalConv1d(residual_channels, 2 * residual_channels, kernel_size, dilation=dilation)
        
        self.condition = nn.Conv1D(cond_channels, 2 * residual_channels, kernel_size=1)
        
        self.residual = nn.Conv1D(residual_channels, residual_channels, kernel_size=1)
        
        self.skip = nn.Conv1D(residual_channels, skip_channels, kernel_size=1)
                

    def _condition(self, x, c, f):
        c = f(c)
        x = x + c
        return x
    
    def forward(self, x, c=None):
        
        x = self.causal(x)
        if c is not None:
            x = self._condition(x, c, self.condition)

        gate, output = x.chunk(2, 1)
        m1= paddle.nn.Sigmoid()   ################
        gate = m1(gate)
        m2= paddle.nn.Tanh()
        output = m2(output)
       
        x = gate * output

        residual = self.residual(x)
        skip = self.skip(x)

        return residual, skip


class WaveNet_x_xy(nn.Layer):

    def __init__(self, blocks=1, layers=10, kernel_size=2, skip_channels=128,
                 residual_channels=128, latent_d=129, shift_input=True):
        super(WaveNet_x_xy, self).__init__()

        self.blocks = blocks
        self.layer_num = layers
        self.kernel_size = kernel_size
        self.skip_channels = skip_channels
        self.residual_channels = residual_channels
        self.cond_channels = latent_d
        self.classes = 256
        self.shift_input = shift_input

        layers = []
        for _ in range(self.blocks):
            for i in range(self.layer_num):
                dilation = 2 ** i
                layers.append(WavenetLayer(self.residual_channels, self.skip_channels, self.cond_channels,
                                           self.kernel_size, dilation))
        self.layers = nn.LayerList(layers)

        self.first_conv = CausalConv1d(1, self.residual_channels, kernel_size=self.kernel_size)
        
        self.skip_conv = nn.Conv1D(self.residual_channels, self.skip_channels, kernel_size=1)

        self.condition = nn.Conv1D(self.cond_channels, self.skip_channels, kernel_size=1)
        
        self.fc = nn.Conv1D(self.skip_channels, self.skip_channels, kernel_size=1)
        
        self.logits = nn.Conv1D(self.skip_channels, self.classes, kernel_size=1)

        
    def _condition(self, x, c, f):
        c = f(c)
        x = x + c
        return x

    def shift_right(self, x):
        #x = F.pad(x, (1, 0))     #torch 实现
        x = F.pad(x,pad=[1,0],data_format='NCL') #paddle 实现
        return x[:, :, :-1]
    
    def forward(self, x, c=None,RB_attr=True):
        x = x.unsqueeze(1)
        x = x / 255.0 - 0.5

        #if self.shift_input:
        #    x = self.shift_right(x)

        residual = self.first_conv(x)
        skip = self.skip_conv(residual)
        #########------------
        #### add RB to net 
        #if( RB_attr ):
        #    for layer in self.layers:  #1  * 10
        #        r, s = layer(residual, c)
        #        residual = residual + r
        #        skip = skip + s
        #####-----------------
        #
        skip = F.relu(skip)
        skip = self.fc(skip)
        
        if c is not None:
            skip = self._condition(skip, c, self.condition)
        skip = F.relu(skip)
        skip = self.logits(skip)

        return skip

if __name__ == '__main__':
    parser = argparse.ArgumentParser()

    parser.add_argument('--rb', action='store_true', help='Enable RB layer')
    
    args = parser.parse_args()
        
    paddle.set_device("gpu:0")
    
    ############
    
    decorder = WaveNet_x_xy(3)
    
    ################
    info = \
    '''
      
      paddle 2.1
      cuda 10.2
    
    Please NOTE: device: 0, GPU Compute Capability: 6.1, 
    Driver API Version: 11.1, Runtime API Version: 10.2
    device: 0, cuDNN Version: 7.6.
    '''
    print( 'env.. ','\r\n',info)
    
    import pickle
    from pathlib import Path
    fp = Path(r'samples.npy')
    
    from pathlib import Path
    fp1 = Path(r'decoder_c.npy') 
 
    ##########################
    samples = paddle.to_tensor(pickle.loads(fp.read_bytes() )  )#paddle.uniform( [2,      16000] )
    cin =     paddle.to_tensor(pickle.loads(fp1.read_bytes() )  )#paddle.uniform( [2, 129, 16000] )
    
    ##optime....
    opt_dec = optim.Adam(parameters =decorder.parameters() ,#不加list() 速度快了一点点
                learning_rate=1e-3)
    
    #---train
    decorder.train()
    
    import time
    start = time.time()
    
    for  ep in    range(1001):
        deco = decorder( samples,c = cin,RB_attr=args.rb)
        loss = F.cross_entropy(deco,paddle.cast(samples, paddle.int64),axis=1).unsqueeze(0).mean()
        if(ep % 100 == 0):
            strs = "Yes" if args.rb else "No"
            #print(f" {strs} RB take {time.time()-start} s        /100it ")
            start = time.time()
            print(loss)
        opt_dec.clear_grad()    
                
        loss.backward()     
        opt_dec.step()    
        samples *=  0.0096
        cin *= 0.0095     
 
```
